npm/pkg/dataplane/ipsets/ipset.go

```diff
@@ -163,47 +175,77 @@ func getSetKind(setType SetType) SetKind {
 	}
 }
 
-func (set *IPSet) AddMemberIPSet(memberIPSet *IPSet) {
-	set.MemberIPSets[memberIPSet.Name] = memberIPSet
-}
-
-func (set *IPSet) IncIpsetReferCount() {
-	set.IpsetReferCount++
+func (set *IPSet) incIPSetReferCount() {
+	set.ipsetReferCount++
 }
 
-func (set *IPSet) DecIpsetReferCount() {
-	if set.IpsetReferCount == 0 {
+func (set *IPSet) decIPSetReferCount() {
+	if set.ipsetReferCount == 0 {
 		return
 	}
-	set.IpsetReferCount--
+	set.ipsetReferCount--
+}
+
+func (set *IPSet) incKernelReferCount() {
+	set.kernelReferCount++
 }
 
-func (set *IPSet) AddSelectorReference(netPolName string) {
-	set.SelectorReference[netPolName] = struct{}{}
+func (set *IPSet) decKernelReferCount() {
+	if set.kernelReferCount == 0 {
+		return
+	}
+	set.kernelReferCount--
 }
 
-func (set *IPSet) DeleteSelectorReference(netPolName string) {
-	delete(set.SelectorReference, netPolName)
+func (set *IPSet) addReference(referenceName string, referenceType ReferenceType) {
+	switch referenceType {
+	case SelectorType:
+		set.SelectorReference[referenceName] = struct{}{}
+	case NetPolType:
+		set.NetPolReference[referenceName] = struct{}{}
+	default:
+		log.Logf("IPSet_addReference: encountered unknown ReferenceType")
+	}
 }
 
-func (set *IPSet) AddNetPolReference(netPolName string) {
-	set.NetPolReference[netPolName] = struct{}{}
+func (set *IPSet) deleteReference(referenceName string, referenceType ReferenceType) {
+	switch referenceType {
+	case SelectorType:
+		delete(set.SelectorReference, referenceName)
+	case NetPolType:
+		delete(set.NetPolReference, referenceName)
+	default:
+		log.Logf("IPSet_deleteReference: encountered unknown ReferenceType")
+	}
 }
 
-func (set *IPSet) DeleteNetPolReference(netPolName string) {
-	delete(set.NetPolReference, netPolName)
+func (set *IPSet) shouldBeInKernel() bool {
+	return set.usedByNetPol() || set.referencedInKernel()
 }
 
-func (set *IPSet) CanBeDeleted() bool {
-	return len(set.SelectorReference) == 0 &&
-		len(set.NetPolReference) == 0 &&
-		set.IpsetReferCount == 0 &&
+func (set *IPSet) canBeDeleted() bool {
+	return !set.usedByNetPol() &&
+		!set.referencedInList() &&
 		len(set.MemberIPSets) == 0 &&
```

Comment:
check for null pointer here for MemberIPSets and IPPodKey in case of List, IPPodKey  may be nil and same with other way around

Reply:
len(nil) is actually 0